core/src/main/java/org/infinispan/xsite/ClusteredCacheBackupReceiver.java

```diff
@@ -200,13 +202,20 @@ private static PrivateMetadata internalMetadata(IracMetadata metadata) {
    }
 
    @Override
-   public final CompletionStage<Void> handleRemoteCommand(VisitableCommand command, boolean preserveOrder) {
+   public final CompletionStage<Object> handleRemoteCommand(ReplicableCommand command, boolean preserveOrder) {
+      //currently, it only handles sync xsite requests.
+      //async xsite requests are handle by the other methods.
+      assert !preserveOrder;
       try {
-         //currently, it only handles sync xsite requests.
-         //async xsite requests are handle by the other methods.
-         assert !preserveOrder;
-         //noinspection unchecked
-         return (CompletableFuture<Void>) command.acceptVisitor(null, defaultHandler);
+         if (command instanceof VisitableCommand) {
+            //noinspection unchecked
+            return (CompletionStage<Object>) ((VisitableCommand)command).acceptVisitor(null, defaultHandler);
+         } else if (command instanceof CacheRpcCommand) {
```

Comment:
-1 `CacheRpcCommand` doesn't require anything from the `BackupReceiver` and you have the `ComponentRegistry` available in `GlobalInboundInvocationHandler`

I would change `XSiteReplicateCommand` to the following and make the if-else in `SingleXSiteRpcCommand`
```
   public CompletionStage<O> performInLocalSite(ComponentRegistry registry, boolean preserveOrder) {
      //by default, the command is executed against BackupReceiver.
      return execute(registry.getBackupReceiver().running(), preserveOrder);
   }
   //abstract? or not?
   public abstract CompletionStage<O> execute(BackupReceiver receiver, boolean preserveOrder);
```

Reply:
Okay, I will undo the ClusteredCacheBackupReceiver part and change it up a bit more.